### PR TITLE
GDScript: Fix lambda resolution with cyclic references

### LINKED ISSUE #70592: Cannot access member variable from lambda body due cyclic reference
### Godot version

4.0.beta10

### System information

Windows 10

### Issue description

For top-level class variables, a lambda stored in a dictionary cannot reference the dictionary members.

### Steps to reproduce

Assuming `MyAutoload.gd` is an autoload class, this code that worked in beta 8, no longer works in beta 10.

```gdscript
var f = func():
	MyAutoload.MyDict._str = "New string."

var MyDict: Dictionary = {
	"_str": "",
	"fun": f
}

# Parser Error: Could not resolve member "_OtherClass".
```

Autoload is unrelated however, just wanted to demonstrate how there's an inconsistent error if this is intended behavior. The following occurs regardless of autoload. If the "MyAutoload" part is removed from the lambda, a "cyclic reference" error occurs.
```gdscript
var f = func():
	MyDict._str = "New string."

var MyDict: Dictionary = {
	"_str": "",
	"fun": f
}

# Parser Error: Could not resolve member "f": Cyclic reference
```

### Minimal reproduction project

[Issue_70592_BugProject.zip](https://github.com/godotengine/godot/files/10304077/Issue_70592_BugProject.zip)

### PR DESCRIPTION
**Pull request by dalexeev**
* Closes #70592.
* This PR allows lambda **body** to use class members in any order (but you still have to consider the order in which initializer expressions are evaluated if they call lambdas).
* The lambda signature is still respected in member resolution (because argument initializers don't have to be literals).
* `call()` arguments should also not create cyclic references, as they are part of the outer expression, not part of lambda body.

### CODE REVIEW

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -5288,11 +5277,51 @@ void GDScriptAnalyzer::downgrade_node_type_source(GDScriptParser::Node *p_node)
...
 void GDScriptAnalyzer::mark_lambda_use_self() {
-	for (GDScriptParser::LambdaNode *lambda : lambda_stack) {
+	GDScriptParser::LambdaNode *lambda = current_lambda;
+	while (lambda != nullptr) {
 		lambda->use_self = true;
+		lambda = lambda->parent_lambda;
 	}
 }
 
+void GDScriptAnalyzer::resolve_pending_lambda_bodies() {
```
Comment (dalexeev) [inline modules/gdscript/gdscript_analyzer.cpp:5287]:
```suggestion
void GDScriptAnalyzer::resolve_pending_lambda_bodies() {
	if (pending_body_resolution_lambdas.is_empty()) {
		return;
	}

```

**Review by adamscott — APPROVED** [review]
Everything seems in order!

File: modules/gdscript/gdscript_analyzer.cpp (hunk 1 above)
Comment (YuriSizov) [reply modules/gdscript/gdscript_analyzer.cpp:5287]: Do you want to do anything with this? It looks sensible.

Comment (adamscott) [reply modules/gdscript/gdscript_analyzer.cpp:5287]: Did you test did, Danil?

Comment (YuriSizov) [reply modules/gdscript/gdscript_analyzer.cpp:5287]: I mean, it just avoids moving references around and clearing an already empty container.

**Comment by YuriSizov** [comment]
Thanks!

**Comment by rune-scape** [comment]
im curious if the lambda body resolution could be delayed until the very end of the current scope, (class, function, script) without any issues

**Comment by dalexeev** [comment]
Lambda body resolution is delayed until the end of the member variable definition (otherwise you cannot call the lambda recursively) and until the end of the statement in the case of local expressions, not until the end of the class/function. No problems found so far, let me know if you find one.

**Pull request merged**